File: app/schemas/strategy_dry_run_auto_buy.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


StrategyProfileName = Literal["safe", "balanced", "aggressive"]


class ProfileAwareDryRunAutoBuyRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    provider: str = "kis"
    market: str = "KR"
    # ``profile_name`` is the legacy strategy/risk preset identity. Custom
    # automation profile identity must travel separately.
    profile_name: StrategyProfileName | None = None
    automation_profile_key: str | None = Field(default=None, max_length=80)
    automation_profile_name: str | None = Field(default=None, max_length=120)
    symbol: str | None = None
    max_candidates: int = Field(default=5, ge=1, le=20)
    trigger_source: str = Field(
        default="manual",
        min_length=1,
        max_length=80,
    )
    use_watchlist: bool = True
    save_logs: bool = True

    @field_validator("symbol")
    @classmethod
    def normalize_symbol(cls, value: str | None) -> str | None:
        if value is None:
            return None
        symbol = str(value).strip().upper()
        if not symbol:
            return None
        if not symbol.isdigit() or len(symbol) > 6:
            raise ValueError("KIS symbol must be numeric.")
        return symbol.zfill(6)

    @field_validator("automation_profile_key", "automation_profile_name")
    @classmethod
    def normalize_automation_profile_context(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = str(value).strip()
        return normalized or None


class ProfileAwareDryRunAutoBuyResponse(BaseModel):
    status: str
    action: str
    provider: str
    market: str
    active_profile: str
    profile_key: str | None = None
    profile_name: str | None = None
    automation_profile_key: str | None = None
    automation_profile_name: str | None = None
    legacy_profile_name: StrategyProfileName | None = None
    profile_provider: str | None = None
    profile_market: str | None = None
    selected_symbol: str | None = None
    selected_symbol_name: str | None = None
    candidate_count: int
    configured_symbol_count: int = 0
    analyzed_symbol_count: int = 0
    quant_candidate_count: int = 0
    quant_scored_count: int = 0
    gpt_candidate_count: int = 0
    gpt_target_count: int = 0
    gpt_completed_count: int = 0
    gpt_failed_count: int = 0
    gpt_not_run_count: int = 0
    final_candidate_count: int = 0
    final_ranked_count: int = 0
    profile_eligible_symbol_count: int = 0
    profile_price_filtered_count: int = 0
    execution_candidate_count: int = 0
    profile_exclusion_counts: dict[str, int] = Field(default_factory=dict)
    preview_status: str = "unknown"
    preview_error: str | None = None
    candidates: list[dict[str, Any]] = Field(default_factory=list)
    buy_score: float | None = None
    final_buy_score: float | None = None
    sell_score: float | None = None
    final_score: float | None = None
    selected_quant_buy_score: float | None = None
    selected_quant_sell_score: float | None = None
    selected_ai_buy_score: float | None = None
    selected_ai_sell_score: float | None = None
    selected_gpt_analysis_status: str | None = None
    selected_gpt_used: bool = False
    selected_gpt_reason: str | None = None
    selected_final_buy_score: float | None = None
    selected_final_sell_score: float | None = None
    selected_confidence: float | None = None
    selected_candidate_observability: dict[str, Any] = Field(default_factory=dict)
    required_entry_score: float = 0
    confidence: float | None = None
    target_risk_approved: bool
    target_risk_result: dict[str, Any] = Field(default_factory=dict)
    recommended_notional_krw: float
    recommended_notional_pct: float
    simulated_quantity: int
    simulated_price: float | None = None
    simulated_notional_krw: float
    reason: str
    risk_flags: list[str] = Field(default_factory=list)
    gating_notes: list[str] = Field(default_factory=list)
    signal_id: int | None = None
    trade_run_id: int | None = None
    simulated_order_id: int | None = None
    data_quality: dict[str, Any] = Field(default_factory=dict)
    data_quality_limited: bool = False
    data_quality_notes: list[str] = Field(default_factory=list)
    data_quality_reduction_reasons: list[str] = Field(default_factory=list)
    sizing_mode: str = "equity_pct"
    fixed_budget_krw: float = 0.0
    target_position_pct: float = 0.0
    available_cash_krw: float | None = None
    total_assets_krw: float | None = None
    configured_max_order_notional_krw: float = 0.0
    hard_max_order_notional_krw: float = 1_000_000.0
    base_order_cap_krw: float = 0.0
    effective_max_order_notional_krw: float = 0.0
    order_cap_source: str = "equity_pct"
    safety: dict[str, Any] = Field(default_factory=dict)
    created_at: str | None = None


class ProfileAwareDryRunRecentResponse(BaseModel):
    provider: str
    market: str
    count: int
    items: list[dict[str, Any]] = Field(default_factory=list)
    safety: dict[str, Any] = Field(default_factory=dict)


class ProfileAwareDryRunSummaryResponse(BaseModel):
    provider: str
    market: str
    today: dict[str, Any] = Field(default_factory=dict)
    month: dict[str, Any] = Field(default_factory=dict)
    profiles: dict[str, Any] = Field(default_factory=dict)
    safety: dict[str, Any] = Field(default_factory=dict)

File: lib/models/strategy_dry_run_auto_buy.dart
class StrategyDryRunAutoBuyResult {
  const StrategyDryRunAutoBuyResult({
    required this.status,
    required this.action,
    required this.provider,
    required this.market,
    required this.activeProfile,
    required this.candidateCount,
    required this.candidates,
    required this.targetRiskApproved,
    required this.targetRiskResult,
    required this.recommendedNotionalKrw,
    required this.recommendedNotionalPct,
    required this.simulatedQuantity,
    required this.simulatedNotionalKrw,
    required this.reason,
    required this.riskFlags,
    required this.gatingNotes,
    required this.dataQuality,
    required this.safety,
    this.selectedSymbol,
    this.selectedSymbolName,
    this.buyScore,
    this.sellScore,
    this.finalScore,
    this.confidence,
    this.simulatedPrice,
    this.signalId,
    this.tradeRunId,
    this.simulatedOrderId,
    this.createdAt,
  });

  final String status;
  final String action;
  final String provider;
  final String market;
  final String activeProfile;
  final String? selectedSymbol;
  final String? selectedSymbolName;
  final int candidateCount;
  final List<Map<String, dynamic>> candidates;
  final double? buyScore;
  final double? sellScore;
  final double? finalScore;
  final double? confidence;
  final bool targetRiskApproved;
  final Map<String, dynamic> targetRiskResult;
  final double recommendedNotionalKrw;
  final double recommendedNotionalPct;
  final int simulatedQuantity;
  final double? simulatedPrice;
  final double simulatedNotionalKrw;
  final String reason;
  final List<String> riskFlags;
  final List<String> gatingNotes;
  final int? signalId;
  final int? tradeRunId;
  final int? simulatedOrderId;
  final Map<String, dynamic> dataQuality;
  final Map<String, dynamic> safety;
  final DateTime? createdAt;

  bool get wouldBuy => action == 'would_buy';
  bool get blocked => action == 'blocked';

  factory StrategyDryRunAutoBuyResult.fromJson(Map<String, dynamic> json) {
    return StrategyDryRunAutoBuyResult(
      status: _string(json['status'], 'ok'),
      action: _string(json['action'], 'hold'),
      provider: _string(json['provider'], 'kis'),
      market: _string(json['market'], 'KR'),
      activeProfile: _string(json['active_profile'], 'safe'),
      selectedSymbol: _nullableString(json['selected_symbol']),
      selectedSymbolName: _nullableString(json['selected_symbol_name']),
      candidateCount: _int(json['candidate_count']),
      candidates: _maps(json['candidates']),
      buyScore: _nullableDouble(json['buy_score']),
      sellScore: _nullableDouble(json['sell_score']),
      finalScore: _nullableDouble(json['final_score']),
      confidence: _nullableDouble(json['confidence']),
      targetRiskApproved: json['target_risk_approved'] == true,
      targetRiskResult: _map(json['target_risk_result']),
      recommendedNotionalKrw: _double(json['recommended_notional_krw']),
      recommendedNotionalPct: _double(json['recommended_notional_pct']),
      simulatedQuantity: _int(json['simulated_quantity']),
      simulatedPrice: _nullableDouble(json['simulated_price']),
      simulatedNotionalKrw: _double(json['simulated_notional_krw']),
      reason: _string(json['reason'], ''),
      riskFlags: _strings(json['risk_flags']),
      gatingNotes: _strings(json['gating_notes']),
      signalId: _nullableInt(json['signal_id']),
      tradeRunId: _nullableInt(json['trade_run_id']),
      simulatedOrderId: _nullableInt(json['simulated_order_id']),
      dataQuality: _map(json['data_quality']),
      safety: _map(json['safety']),
      createdAt: _dateTime(json['created_at']),
    );
  }
}

class StrategyDryRunAutoBuyRecent {
  const StrategyDryRunAutoBuyRecent({
    required this.provider,
    required this.market,
    required this.items,
    required this.safety,
  });

  final String provider;
  final String market;
  final List<StrategyDryRunAutoBuyResult> items;
  final Map<String, dynamic> safety;

  StrategyDryRunAutoBuyResult? get latest => items.isEmpty ? null : items.first;

  factory StrategyDryRunAutoBuyRecent.fromJson(Map<String, dynamic> json) {
    final rawItems = json['items'];
    return StrategyDryRunAutoBuyRecent(
      provider: _string(json['provider'], 'kis'),
      market: _string(json['market'], 'KR'),
      items: rawItems is List
          ? [
              for (final item in rawItems)
                if (item is Map)
                  StrategyDryRunAutoBuyResult.fromJson(
                    Map<String, dynamic>.from(item),
                  ),
            ]
          : const [],
      safety: _map(json['safety']),
    );
  }
}

String _string(Object? value, String fallback) {
  final text = value?.toString().trim();
  return text == null || text.isEmpty ? fallback : text;
}

String? _nullableString(Object? value) {
  final text = value?.toString().trim();
  return text == null || text.isEmpty || text == 'null' ? null : text;
}

double _double(Object? value) => _nullableDouble(value) ?? 0;

double? _nullableDouble(Object? value) {
  if (value == null) return null;
  if (value is num) return value.toDouble();
  return double.tryParse(value.toString().replaceAll(',', '').trim());
}

int _int(Object? value) => _nullableInt(value) ?? 0;

int? _nullableInt(Object? value) {
  if (value == null) return null;
  if (value is num) return value.toInt();
  return int.tryParse(value.toString().trim());
}

List<String> _strings(Object? value) {
  if (value is! List) return const [];
  return [
    for (final item in value)
      if (item?.toString().trim().isNotEmpty == true) item.toString().trim(),
  ];
}

List<Map<String, dynamic>> _maps(Object? value) {
  if (value is! List) return const [];
  return [
    for (final item in value)
      if (item is Map) Map<String, dynamic>.from(item),
  ];
}

Map<String, dynamic> _map(Object? value) {
  return value is Map ? Map<String, dynamic>.from(value) : const {};
}

DateTime? _dateTime(Object? value) {
  final text = value?.toString().trim();
  return text == null || text.isEmpty ? null : DateTime.tryParse(text);
}
